[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.14536572.svg)](https://doi.org/10.5281/zenodo.14536572)
## Overview
This repository contains the jupyter lab notebooks of analysis and plotting after performing pySpade analysis of breast cancer associated enhancers perturbations.

[Link to pySpade](https://github.com/yihan1119/pySpade)

![Over-view](./MISC/Overview1.png "Overview")

## Requirement
* Python (3.7 +)
* Numpy (1.16 +)
* Pandas (0.25 +)
* Scipy (1.1 +)
* Matplotlib (3.3 +)
* pyCircos (0.3.0 +)

## Fastq files
* [GSE223812]: Bulk ATAC-seq of breast cancer cell lines. (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE223812)
* [GSE224986]: Single cell libraries. (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE224986)
* [GSE228332]: Bulk RNA-seq validation experiments. (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE228332)

## How to cite 
Wang, Y. et al. Enhancer regulatory networks globally connect non-coding breast cancer loci to cancer genes. Genome Biol 26, 10 (2025).
[Link to paper](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-025-03474-0)

## Contacts
* Yihan Wang `Yihan.Wang@UTSouthwestern.edu`
* Gary Hon `Gary.Hon@UTSouthwestern.edu`
